
<div class="overview-container">
  <div class="container">
    <div class="row">
      
      <div class="col-md-12">
        <div class="title">
          <h3>Overview</h3>
        </div>
        
        <div class="description">
          <p>Exposure to environmental chemicals is known to play a significant role in multiple adverse effects, including carcinogenesis (carcinogens), genotoxicity (genotoxicants), and metabolic disruption (metabolism disrupting chemicals [MDCs]). While for some chemicals their adverse effects are well-known, only a small fraction of those in commercial use and to which the human population is exposed has undergone thorough testing. Furthermore, many of the biological mechanisms of actions of chemical carcinogens and MDCs are poorly understood. For example, the current gold standard for chemical carcinogenicity testing is the 2-year rat bioassay, which is expensive and time-consuming, and its relevance to human carcinogenesis has been questioned. To date, about ~1500 chemicals have been tested for long-term carcinogenicity in rodent models by this approach, out of the ~85,000 chemicals used in common household products or industrial settings. Here, we aim to collect toxicogenomics screens profiling the short-term chemical exposures in mammalian cell lines. This collection will support the investigation of candidate biomarkers (genes and pathways) of chemical exposure, and will support the investigation of the chemicals' mechanisms of action.</p>
        </div>
      </div>

      <div class="col-md-12">
        <div class="title">
          <h3>Datasets</h3>
        </div>
        
        <div class="header-1">
          <h5>Carcinogenicity Screens</h5>
        </div>
        
        <div class="description">
          <p>Currently, we have profiled two sets of chemicals in projects aimed at predicting liver (Li et al., 2019) and breast carcinogenicity, respectively, using short-term transcriptomics assays based on human cell lines carcinogenome.org. Our previous studies (Gusenleitner et al., 2014) have shown that carcinogenicity prediction is exquisitely tissue-specific, and requires the fine tuning of the predictive model to the tissue being tested, hence the need for generating tissue-specific screens. The first set of chemicals is composed of known liver carcinogens (according to evidence of tumor growth in rat livers according to the CPDB and additional sources), non-carcinogens (known not to induce tumors in any sites in rats according to the CPDB), and other chemicals of interest to our collaborators in the Boston University Superfund Research Program. HEPG2 (human liver cancer) cell lines are exposed to each of the chemicals at 6 different doses for 24H. The second set of chemicals is composed of known or suspected breast carcinogens (according to the CPDB or (Rudel et. al., 2007)), known lung carcinogens, non-carcinogens, or collaborator-selected chemicals. MCF10A (human non-transformed breast epithelial) cell lines as well as MCF10A with P53 knock-down, are exposed to each of the chemicals at 3 different doses for 24 hours and their gene expression profiled.</p>
        </div>
        
        <div class="header-1">
          <h5>Metabolism Disrupting Chemical (MDC) Screen</h5>
        </div>
        
        <div class="description">
          <p>We have also profiled a set of chemicals known or suspected to play a role in metabolism regulation. 3T3-L1 cells were differentiated in the presence of 76 chemicals (negative controls, synthetic nuclear receptor ligands known to influence adipocyte biology, suspected environmental PPARÎ³ ligands), and their mRNA expression was quantified by highly multiplexed RNA-Seq (Kim, Reed, et al., 2020).</p>
        </div>
        
        <div class="header-1">
          <h5>Public Toxicogenomics Databases</h5>
        </div>
          
        <div class="description">
          <p>In addition to in house-generated data, we included high-profile toxicogenomics screens previously generated. Currently, we are hosting the NIEHS/NTP-supported DrugMatrix, and the NBIOHN-supported TG-GATEs. We plan to incorporate additional resources going forward.</p>
        </div>
      </div>
      
    </div>
  </div>
</div>
  